README.md
# Redinn Core

## The ulitmate service for managing enterprises

### packages

- backend - a REST/Graphql api
- frontend - a universal app


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2021-07-23)


### Bug Fixes

* **api:** :rotating_light: ts issues ([964c970](https://github.com/Redinn-Labs/redinn-core/commit/964c9702784fccd3ee039cdc1c24908dc9e32cd9))
* **auth:** :bug: toke field ([6ccb7ee](https://github.com/Redinn-Labs/redinn-core/commit/6ccb7ee19c34848b7a5fed9a37f220355cbd3d9f))
* **graphql:** :bug: addenterprise Mutation ([688ecaf](https://github.com/Redinn-Labs/redinn-core/commit/688ecaf6ed44cfd417fb1a22aa42fafec713471e))
* **graphql:** :bug: deleteEnterprise errors ([52aaf66](https://github.com/Redinn-Labs/redinn-core/commit/52aaf666bcaa56113c629e3df5793967775d3fd0))
* **graphql:** :goal_net: error status codes ([afe486e](https://github.com/Redinn-Labs/redinn-core/commit/afe486efb5a2e7d06343b61c5855000fd43a399f))
* **media:** simplify routing in the photo uploader ([9495d3e](https://github.com/Redinn-Labs/redinn-core/commit/9495d3ec7152b4a63dad0e3f6375b7745f870db6))
* **website:** :bug: addEnterprise flow ([7b73917](https://github.com/Redinn-Labs/redinn-core/commit/7b73917f48496c41ed72965fbd36fc18587d3c1f))
* **website:** :bug: conditions for fetching enterprises ([55332d6](https://github.com/Redinn-Labs/redinn-core/commit/55332d61bd5f4301d3607226acc058204bbfd35c))
* **website:** :rotating_light: linters' warnings ([e8e2889](https://github.com/Redinn-Labs/redinn-core/commit/e8e288959aac9fc2f308403f5c0ca650bebdf9eb))
* **website:** :truck: addbusiness -> addenterprise ([bb807ec](https://github.com/Redinn-Labs/redinn-core/commit/bb807ecc212b0a098504d078b21faade5bc7d40a))
* :art: enterprise middleware ([e6e5daa](https://github.com/Redinn-Labs/redinn-core/commit/e6e5daad0e573fcf103e3a4a665de43903f3d003))
* impove performace - remove unnecessary css ([e7d7133](https://github.com/Redinn-Labs/redinn-core/commit/e7d7133a9923aef0be92b228e707a7211758e251))


### Features

* **api:** :building_construction: db package ([86606d3](https://github.com/Redinn-Labs/redinn-core/commit/86606d36f3ab757357fb1f767f6056e8b485d686))
* **api:** :goal_net: better error messages ([3437805](https://github.com/Redinn-Labs/redinn-core/commit/343780580e2e385d5e7b776d7622bc9e03af23ee))
* **api:** :lock: route for registering ([3962fb7](https://github.com/Redinn-Labs/redinn-core/commit/3962fb7431e8743c803558b922cf29b0df9ddf8f))
* **api:** :sparkles: graphql playground ([55a7715](https://github.com/Redinn-Labs/redinn-core/commit/55a7715fc792b28fd4bf42d3109a836b44ef76c3))
* **api:** :sparkles: types for users ([3a7b3fb](https://github.com/Redinn-Labs/redinn-core/commit/3a7b3fbf01ccb2d8deb1b96fb20c2a12aeb4d271))
* **auth:** :lock: /login route ([a9da29a](https://github.com/Redinn-Labs/redinn-core/commit/a9da29a94e0970694c20366b2a3f66d1150b263a))
* **auth:** :lock: bcrypt for passwords ([2a1cdb4](https://github.com/Redinn-Labs/redinn-core/commit/2a1cdb4fea48b78baf7ce367944fb07f3ddcbb3c))
* **auth:** :sparkles: add basic jwt + login authorization and user registration ([7682d31](https://github.com/Redinn-Labs/redinn-core/commit/7682d313c09d5d419c9404392b64c0e21e2105c4))
* **core:** :art: simplify the experience of adding an enterprise ([66f5b04](https://github.com/Redinn-Labs/redinn-core/commit/66f5b04630257e235cded30c5ac3dd31f806873f))
* **core:** :lock: add middleware that redirects a user to /addbusiness if they don't have any enterprises ([5694310](https://github.com/Redinn-Labs/redinn-core/commit/569431007ba4363bcc0b4ee05106605098fd30e1))
* **core:** :sparkles: enable users to add enterprises ([4d158bf](https://github.com/Redinn-Labs/redinn-core/commit/4d158bf06f25c0830608237a00e135010eb0822b))
* **graphql:** :sparkles: add auth to graphql and allow user's data fetching ([88bd05f](https://github.com/Redinn-Labs/redinn-core/commit/88bd05fd82f7e10d26966e84b1ff37c9e9d2e61d))
* **graphql:** :sparkles: add mutation - delete enterprise ([9dead37](https://github.com/Redinn-Labs/redinn-core/commit/9dead371d619b3fe308393991c649166badd4190))
* **graphql:** :sparkles: add types and model for the enterprise data type ([e8cd5b6](https://github.com/Redinn-Labs/redinn-core/commit/e8cd5b6d12238dde63d2dbfa471d3ddbe9d41daf))
* **graphql:** :sparkles: begin writing types and resolvers ([535ec03](https://github.com/Redinn-Labs/redinn-core/commit/535ec03f5e5d71904b3d564ec729dcfedde5cd70))
* **graphql:** :sparkles: enterprise resolvers ([5ec8f34](https://github.com/Redinn-Labs/redinn-core/commit/5ec8f34fefff12ea966aa48697ae57b54b53e297))
* **graphql:** :sparkles: fetch users from the db ([5471bf6](https://github.com/Redinn-Labs/redinn-core/commit/5471bf635e2394517cbcd8a04a728fc1400ce2fe))
* **graphql:** :zap: utils ([bf3d484](https://github.com/Redinn-Labs/redinn-core/commit/bf3d484c5689cdaabb1d3f2d3240a725690ac82b))
* :heavy_plus_sign: load .env variables thanks to dotenv ([55d3d16](https://github.com/Redinn-Labs/redinn-core/commit/55d3d164c948ae76365daef15374393b8fde4de5))
* :loud_sound: add morgan for better logs ([a562174](https://github.com/Redinn-Labs/redinn-core/commit/a562174129eb725f74d164cc730b920dbb7b6be2))
* :sparkles: add a db model for enterprises ([e55e3c9](https://github.com/Redinn-Labs/redinn-core/commit/e55e3c90350dabb3258ede1d8e13aae951b2633a))
* :sparkles: add authentication, authorization and registration ([d3cfdd4](https://github.com/Redinn-Labs/redinn-core/commit/d3cfdd4079120acd4b07895458f9cbac117bb485))
* :sparkles: add login functionality thanks to Nuxt Auth ([7ba46d6](https://github.com/Redinn-Labs/redinn-core/commit/7ba46d6b481730cabedcfee035073f5a2dfbed05))
* :wrench: add types for process.env variables ([c3cb99c](https://github.com/Redinn-Labs/redinn-core/commit/c3cb99c07f0adf0ec42c7b6a991d2f8fb62eabb9))
* add types for process.env variables ([1b4e660](https://github.com/Redinn-Labs/redinn-core/commit/1b4e660d3922f2c9b95da9b8fd6e416f7dd5429c))

utils/test.go
package main

import (
	"fmt"
	"log"
	"net/http"
	"net/http/httputil"
	"net/url"
	"os"
	"os/exec"
	"syscall"
	"time"
)

var RESET = "\033[0m"
var RED = "\033[31m"
var GREEN = "\033[32m"

func main() {
	// get the current working directory
	CWD, cwderr := os.Getwd()
	if cwderr != nil {
		log.Fatal(cwderr)
	}
	DBSTORAGE := CWD + "/test_db"
	// create a temporary folder for mongodb
	// if already exists, clear content
	if _, staterr := os.Stat(DBSTORAGE); !os.IsNotExist(staterr) {
		if clearerr := os.RemoveAll(DBSTORAGE); clearerr != nil {
			log.Fatal(clearerr)
		}
	}
	if Mkdirerr := os.Mkdir(DBSTORAGE, 0755); Mkdirerr != nil {
		log.Fatal(Mkdirerr)
	}
	// start a mongodb instance
	fmt.Println(GREEN + "preparing to run mongodb..." + RESET)

	args := []string{
		"--bind_ip", "127.0.0.1",
		"--port", os.Getenv("DB_PORT"),
		"--dbpath", DBSTORAGE,
	}
	server := exec.Command(os.Getenv("TEST_DB_BIN_PATH"), args...)
	if mongoerr := server.Start(); mongoerr != nil {
		log.Fatal(mongoerr)
	}
	fmt.Println(GREEN + "mongodb is running!" + RESET)

	serversShutdown := setupServers()
	command := exec.Command("lerna", []string{"run", "test"}...)

	fmt.Println(GREEN + "output:\n" + RESET)
	command.Stdout = os.Stdout
	command.Stderr = os.Stderr
	err := command.Run()
	if err != nil {
		fmt.Println(RED+"errors:\n"+RESET, err.Error())
	}
	// shutdown
	fmt.Println(GREEN + "shutting servers down..." + RESET)
	serversShutdown()
	server.Process.Kill()
}

func generateProxy(proxyurl string) *handler {
	url, err := url.Parse(proxyurl)
	if err != nil {
		panic(err)
	}

	director := func(req *http.Request) {
		req.URL.Scheme = url.Scheme
		req.URL.Host = url.Host
	}

	reverseProxy := &httputil.ReverseProxy{Director: director}
	return &handler{proxy: reverseProxy}
}

func setupServers() func() {
	fmt.Println(GREEN + "building frontend & backend..." + RESET)
	// build packages
	build := exec.Command("lerna", []string{"run", "build"}...)
	buildOutput, buildErr := build.CombinedOutput()
	fmt.Println(string(buildOutput))
	if buildErr != nil {
		log.Fatal(buildErr)
	}
	// start servers in packages
	start := exec.Command("lerna", []string{"run", "start"}...)
	start.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}

	start.Stdout = os.Stdout
	start.Stderr = os.Stderr
	if err := start.Start(); err != nil {
		log.Fatal(err)
	}

	fmt.Println(GREEN + "frontend & backend are running!" + RESET)
	fmt.Println(GREEN + "waiting 3 seconds for both ends to warm up..." + RESET)
	time.Sleep(time.Second * 3)

	// setup the proxies
	fmt.Println(GREEN + "setting up the reverse proxy..." + RESET)
	http.Handle("/app/", generateProxy("http://localhost:3000"))
	http.Handle("/api/", generateProxy("http://localhost:3001"))
	go http.ListenAndServe("localhost:3002", nil)
	fmt.Println(GREEN + "reverse proxy is running!" + RESET)

	return func() {
		pgid, err := syscall.Getpgid(start.Process.Pid)
		if err == nil {
			syscall.Kill(-pgid, 15) // note the minus sign
		}
		start.Wait()
	}
}

type handler struct {
	proxy *httputil.ReverseProxy
}

func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	h.proxy.ServeHTTP(w, r)
}


utils/proxy.go
package main

import (
	"fmt"
	"net/http"
	"net/http/httputil"
	"net/url"
)

var RESET = "\033[0m"
var GREEN = "\033[32m"

func generateProxy(proxyurl string) *handler {
	url, err := url.Parse(proxyurl)
	if err != nil {
		panic(err)
	}

	director := func(req *http.Request) {
		req.URL.Scheme = url.Scheme
		req.URL.Host = url.Host
	}

	reverseProxy := &httputil.ReverseProxy{Director: director}
	return &handler{proxy: reverseProxy}
}

func main() {
	// setup the proxies
	http.Handle("/app/", generateProxy("http://localhost:3000"))
	http.Handle("/api/", generateProxy("http://localhost:3001"))
	fmt.Println(GREEN + "reverse proxy is running!" + RESET)
	http.ListenAndServe("localhost:3002", nil)
}

type handler struct {
	proxy *httputil.ReverseProxy
}

func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	h.proxy.ServeHTTP(w, r)
}
